userspace/async/async_metadata_source.tpp

```diff
@@ -0,0 +1,286 @@
+/*
+Copyright (C) 2018 Sysdig, Inc.
+
+This file is part of sysdig.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+#include "logger.h"
+
+#include <assert.h>
+#include <algorithm>
+#include <chrono>
+#include <iostream>
+#include <list>
+#include <string>
+
+namespace sysdig
+{
+
+template<typename key_type, typename metadata_type>
+async_metadata_source<key_type, metadata_type>::async_metadata_source(
+		const uint64_t max_wait_ms,
+		const uint64_t ttl_ms):
+	m_max_wait_ms(max_wait_ms),
+	m_ttl_ms(ttl_ms),
+	m_thread(),
+	m_running(false),
+	m_terminate(false),
+	m_mutex(),
+	m_queue_not_empty_condition(),
+	m_metadata_map()
+{ }
+
+template<typename key_type, typename metadata_type>
+async_metadata_source<key_type, metadata_type>::~async_metadata_source()
+{
+	try
+	{
+		stop();
+	}
+	catch(...)
+	{
+		g_logger.log(std::string(__FUNCTION__) +
+		             ": Exception in destructor",
+		             sinsp_logger::SEV_ERROR);
+	}
+}
+
+template<typename key_type, typename metadata_type>
+uint64_t async_metadata_source<key_type, metadata_type>::get_max_wait() const
+{
+	return m_max_wait_ms;
+}
+
+template<typename key_type, typename metadata_type>
+uint64_t async_metadata_source<key_type, metadata_type>::get_ttl() const
+{
+	return m_ttl_ms;
+}
+
+template<typename key_type, typename metadata_type>
+void async_metadata_source<key_type, metadata_type>::stop()
+{
+	bool join_needed = false;
+
+	{
+		std::unique_lock<std::mutex> guard(m_mutex);
+
+		if(m_running)
+		{
+			m_terminate = true;
+			join_needed = true;
+
+			// The async thread might be waiting for new events
+			// so wake it up
+			m_queue_not_empty_condition.notify_one();
+		}
+	} // Drop the mutex before join()
+
+	if (join_needed)
+	{
+		m_thread.join();
+
+		// Remove any pointers from the thread to this object
+		// (just to be safe)
+		m_thread = std::thread();
+	}
+}
+
+template<typename key_type, typename metadata_type>
+bool async_metadata_source<key_type, metadata_type>::is_running() const
+{
+	std::lock_guard<std::mutex> guard(m_mutex);
```

Comment:
this lock doesn't get you anything

Reply:
Probably not, other than being consistent with the model without worrying too much about the details.  I can remove it if you prefer.

Reply:
here's the problem I've seen, and you can take for it what you want:

when locks are used over-liberally, people lose context of exactly what the lock is supposed to protect. Since I can't put my finger on what it is supposed to protect, I have to be over-liberal with my locking when I edit this code.....since I'm not sure if it WAS supposed to be protecting something. This exhibited at least twice in prior code bases and then we had to go through a clean-up process to remove all the unnecessary locking.

In this case, I look at the function and try to find the reason we're locking here, it just sends the wrong message to the reader.

Reply:
That's legit.  I'll remove it.